### DISCUSSION TITLE
B-Tree

### DISCUSSION THREAD

**Discussion by jangwooooo** [subak-ba/cs-study#48] [자료 구조]
## B-Tree란
데이터베이스와 파일 시스템에서 널리 사용되는 트리 자료구조의 일종입니다.
이진 트리와는 다르게 하나의 노드에 많은 정보를 가지거나, 두 개 이상의 자식을 가질 수도 있습니다.

<br/>

## 노드 구조
B-트리의 노드는 대개 항목과 자식 포인터의 순서 집합으로 표현된다. 
루트 노드를 제외한 모든 노드는 임의의 수 L, U(L<U)에 대해 최소 L 항목, 최대 U 항목을 가지고 있으며, 최대 U+1개의 자식 포인터를 가지고 있다. 
모든 내부 노드에서, 자식 포인터의 수는 언제나 항목 개수보다 하나가 많다. 
모든 리프 노드가 동일한 높이에 있기 때문에, 노드는 일반적으로 리프 노드인지 내부 노드인지 판별하는 수단을 가질 이유가 없다.

각 내부 노드의 항목은 자식 노드를 나누는 구분 값이다. 
예를 들어, 어떤 내부 노드가 3개의 자식 노드를 가지고 있다면, 그 내부 노드는 항목 a1과 a2를 가지고 있어야 한다. 
가장 왼쪽 자식 노드에 있는 모든 값은 a1보다 작으며, 가운데 자식 노드의 모든 값은 a1와 a2의 사이에 있으며, 가장 오른쪽 자식 노드의 모든 값은 a2보다 크게 된다.

<img width="652" alt="스크린샷 2024-04-17 23 45 14" src="https://github.com/subak-ba/cs-study/assets/108380536/adc6a561-d855-49bc-8a0b-03e188c1040f">  

<br/>
<br/>
<br/>

## 사용처
하나의 노드에 여러 자료를 배치하게 되면서 이진 트리보다 훨씬 많은 데이터를 더 효율적으로 저장소에 담을 수 있게 되었습니다.

하드디스크나 SSD와 같은 외부 기억장치는 블럭단위로 파일을 입출력합니다. 
이때 발생하는 입출력의 비용은 파일의 크기와는 상관 없이 동일합니다. 
입출력에 있어서는 1KB짜리 블럭에 1byte짜리 알파벳 하나가 들어가 있어도 1KB가 꽉 차있는 블럭과 차이가 없다는 것이죠.
이 떄 하나의 블럭에 여러 데이터들을 동시에 저장할 수 있다면 블럭을 보다 효율적으로 사용할 수 있겠죠?
그래서 많은 데이터베이스들은 B-Tree를 애용합니다.

직접 B-Tree를 생성하는 것을 눈으로 보고 싶다면 아래 사이트에서 가능합니다
=> https://www.cs.usfca.edu/~galles/visualization/BTree.html

**Comment by onee-only** [comment]
문서 전반적으로 문체가 다양하네요 ㅋㅋ

**Comment by onee-only** [comment]
B-Tree의 연산 관련한 부분도 추가 해 주시면 좋을 것 같아요
